app/src/main/java/de/westnordost/streetcomplete/data/osm/mapdata/MapDataCache.kt

```diff
@@ -404,10 +414,33 @@ class MapDataCache(
             relationIdsByElementKeyCache[node.key]?.let { relationIds.addAll(it) }
             result.put(node, ElementPointGeometry(node.position))
         }
+
+        val nodesToFetch = hashSetOf<Long>()
         for (wayId in wayIds) {
-            result.put(wayCache[wayId]!!, wayGeometryCache[wayId])
-            relationIdsByElementKeyCache[ElementKey(ElementType.WAY, wayId)]?.let { relationIds.addAll(it) }
+            val way = wayCache[wayId]!!
+            val wayGeometry = wayGeometryCache[wayId]
+            result.put(way, wayGeometry)
+            relationIdsByElementKeyCache[way.key]?.let { relationIds.addAll(it) }
+
+            // find all nodes that are part of the way, but not in result
+            if (wayGeometry?.getBounds()?.isCompletelyInside(bbox) == true) continue // no need to check
+            for (nodeId in way.nodeIds) {
+                if (result.getNode(nodeId) != null) continue
+                val cachedNode = spatialCache.get(nodeId) ?: nodeCache[nodeId]
```

Comment:
Why should `spatialCache.get` return not null here?

Reply:
Because the way might have nodes in a cached tile, but outside the fetch bbox